Source/DataModels/Resources/ResourceNavMesh.cpp

```diff
@@ -0,0 +1,1032 @@
+#include "Detour/InputGeom.h"
+#include "DetourCrowd/DetourCrowd.h"
+#include "DetourTileCache/DetourTileCache.h"
+#include "DetourTileCache/DetourTileCacheBuilder.h"
+#include "GL/glew.h"
+#include "Recast/Recast.h"
+#include "Recast/RecastAlloc.h"
+#include "Recast/RecastAssert.h"
+#include "fastlz/fastlz.h"
+
+#define MAX_AGENTS 128
+
+// This value specifies how many layers (or "floors") each navmesh tile is expected to have.
+static const int EXPECTED_LAYERS_PER_TILE = 4;
+
+// These are just sample areas to use consistent values across the samples.
+// The use should specify these base on his needs.
+enum SamplePolyAreas
+{
+	SAMPLE_POLYAREA_GROUND,
+	SAMPLE_POLYAREA_WATER,
+	SAMPLE_POLYAREA_ROAD,
+	SAMPLE_POLYAREA_DOOR,
+	SAMPLE_POLYAREA_GRASS,
+	SAMPLE_POLYAREA_JUMP,
+};
+
+enum SamplePolyFlags
+{
+	SAMPLE_POLYFLAGS_WALK = 0x01,	  // Ability to walk (ground, grass, road)
+	SAMPLE_POLYFLAGS_SWIM = 0x02,	  // Ability to swim (water).
+	SAMPLE_POLYFLAGS_DOOR = 0x04,	  // Ability to move through doors.
+	SAMPLE_POLYFLAGS_JUMP = 0x08,	  // Ability to jump.
+	SAMPLE_POLYFLAGS_DISABLED = 0x10, // Disabled polygon
+	SAMPLE_POLYFLAGS_ALL = 0xffff	  // All abilities.
+};
+
+enum SamplePartitionType
+{
+	SAMPLE_PARTITION_WATERSHED,
+	SAMPLE_PARTITION_MONOTONE,
+	SAMPLE_PARTITION_LAYERS,
+};
+
+static const int MAX_LAYERS = 32;
+
+struct TileCacheData
+{
+	unsigned char* data;
+	int dataSize;
+};
+
+struct FastLZCompressor : public dtTileCacheCompressor
+{
+	virtual int maxCompressedSize(const int bufferSize) override
+	{
+		return (int) (bufferSize * 1.05f);
+	}
+
+	virtual dtStatus compress(const unsigned char* buffer,
+							  const int bufferSize,
+							  unsigned char* compressed,
+							  const int /*maxCompressedSize*/,
+							  int* compressedSize) override
+	{
+		*compressedSize = fastlz_compress((const void* const) buffer, bufferSize, compressed);
+		return DT_SUCCESS;
+	}
+
+	virtual dtStatus decompress(const unsigned char* compressed,
+								const int compressedSize,
+								unsigned char* buffer,
+								const int maxBufferSize,
+								int* bufferSize) override
+	{
+		*bufferSize = fastlz_decompress(compressed, compressedSize, buffer, maxBufferSize);
+		return *bufferSize < 0 ? DT_FAILURE : DT_SUCCESS;
+	}
+};
+
+struct LinearAllocator : public dtTileCacheAlloc
+{
+	unsigned char* buffer;
+	int capacity;
+	int top;
+	int high;
+
+	LinearAllocator(const int cap) : buffer(0), capacity(0), top(0), high(0)
+	{
+		resize(cap);
+	}
+
+	virtual ~LinearAllocator() override
+	{
+		dtFree(buffer);
+	}
+
+	void resize(const int cap)
+	{
+		if (buffer)
+			dtFree(buffer);
+		buffer = (unsigned char*) dtAlloc(cap, DT_ALLOC_PERM);
+		capacity = cap;
+	}
+
+	virtual void reset() override
+	{
+		high = dtMax(high, top);
+		top = 0;
+	}
+
+	virtual void* alloc(const size_t size) override
+	{
+		if (!buffer)
+			return 0;
+		if (top + static_cast<int>(size) > capacity)
+			return 0;
+		unsigned char* mem = &buffer[top];
+		top += size;
```

Comment:
```suggestion
		top += static_cast<int>(size);
```
This raises a warning, try this instead.